Refactoring NEST

tammoippen opened the pull request:
This PR contains our joint effort on refactoring the kernel of NEST. Beside one public API change, the appearance and behavior of NEST should remain unchanged. The refactoring contains:

* The classes `Network`, `Scheduler`, `Communicator` and `ConnectionManager` are marked deprecated and will be removed in the future.
* Their functionality is moved into `Manager` classes:
    * `ConnectionBuilderManager` (will be renamed to `ConnectionManager` later) is responsible for storing, creating and managing connections.
    * `EventDeliveryManager` is responsible for delivering events, both local and remote.
    * `IOManager` is responsible for file IO with the operating system.
    * `LoggingManager` is responsible for creating and distributing logging events (former `Network::message(...)` method).
    * `ModelManager` is responsible for storing node and synapse models and creating new instances from these models.
    * `ModelRangeManager` stores the node model for ranges of nodes.
    * `MPIManager` is responsible for all MPI communication.
    * `MUSICManager` is responsible for all MUSIC related interactions.
    * `NodeManager` is responsible for storing and managing all node objects.
    * `RNGManager` is responsible for creation and access to all random number generators.
    * `SimulationManager` is responsible for starting, stopping and resuming simulations.
    * `SPManager` is responsible for structural plasticity.
    * `VPManager` is responsible for threading and the management of virtual processes.
* The `KernelManager` gives easy access to all `Manager` classes via the function `nest::kernel()`.
* [API change] The `min_delay` and `max_delay` setting and checking is moved from the `ConnectorModel` to the `DelayChecker`s, which are managed by the `ConnectionBuilderManager`. Therefore, the `*_delay` is not set on the default synapse types any more, but during the `SetKernelStatus`.
* Sort the header includes according to the guidelines. Also add the script `extras/include_checker.py` that suggests an include order for a file or the files in a directory.
* Each module contain the beginning for a public NEST API, that can be used to program NEST on the C++ level. The corresponding `SLIModule`s are implemented on top of this API (ongoing work).

I hope I did not miss any changes. I invite you all (@nest/nest-developers) to review and test this PR. @janhahne please see that the gap junctions work as expected. @sdiazpier and @naveau please see that the structural plasticity is working as expected.

gewaltig commented:
Dear Tamo,

This looks like an impressive piece of work! Congratulations already now for taking on this Herculean task of cleaning up the classes.
I’ll definitely have a look at the PR.

Best
Marc-Oliver
On 12 Jan 2016, at 11:02, Tammo Ippen <notifications@github.com<mailto:notifications@github.com>> wrote:


This PR contains our joint effort on refactoring the kernel of NEST. Beside one public API change, the appearance and behavior of NEST should remain unchanged. The refactoring contains:

  *   The classes Network, Scheduler, Communicator and ConnectionManager are marked deprecated and will be removed in the future.
  *   Their functionality is moved into Manager classes:
     *   ConnectionBuilderManager (will be renamed to ConnectionManager later) is responsible for storing, creating and managing connections.
     *   EventDeliveryManager is responsible for delivering events, both local and remote.
     *   IOManager is responsible for file IO with the operating system.
     *   LoggingManager is responsible for creating and distributing logging events (former Network::message(...) method).
     *   ModelManager is responsible for storing node and synapse models and creating new instances from these models.
     *   ModelRangeManager stores the node model for ranges of nodes.
     *   MPIManager is responsible for all MPI communication.
     *   MUSICManager is responsible for all MUSIC related interactions.
     *   NodeManager is responsible for storing and managing all node objects.
     *   RNGManager is responsible for creation and access to all random number generators.
     *   SimulationManager is responsible for starting, stopping and resuming simulations.
     *   SPManager is responsible for structural plasticity.
     *   VPManager is responsible for threading and the management of virtual processes.
  *   The KernelManager gives easy access to all Manager classes via the function nest::kernel().
  *   [API change] The min_delay and max_delay setting and checking is moved from the ConnectorModel to the DelayCheckers, which are managed by the ConnectionBuilderManager. Therefore, the *_delay is not set on the default synapse types any more, but during the SetKernelStatus.
  *   Sort the header includes according to the guidelines. Also add the script extras/include_checker.py that suggests an include order for a file or the files in a directory.
  *   Each module contain the beginning for a public NEST API, that can be used to program NEST on the C++ level. The corresponding SLIModules are implemented on top of this API (ongoing work).

I hope I did not miss any changes. I invite you all (@nest/nest-developers<https://github.com/orgs/nest/teams/nest-developers>) to review and test this PR. @janhahne<https://github.com/janhahne> please see that the gap junctions work as expected. @sdiazpier<https://github.com/sdiazpier> and @naveau<https://github.com/naveau> please see that the structural plasticity is working as expected.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/nest/nest-simulator/pull/205

Commit Summary

  *   Merge branch 'master' into refactoring_nest
  *   Merge branch 'master' into refactoring_nest
  *   Remove unused/ not implemented variables and methods.
  *   Merge branch 'master' into refactoring_nest
  *   Initial Manager commit
  *   Initial ModelManager commit
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into modelmanager
  *   Add def and updateValue for Dictionary objects
  *   Update KernelManager
  *   add further functions in VPManager and replace their occurance in other files
  *   Register ModelManager in Makefile.am and fix comments
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into modelmanager
  *   Finish VPManager; make installcheck successful
  *   Create logging facility for nest.
  *   rename nest.h by nest_types.h
  *   Initial commit for the NEST API
  *   Replace Network::get_network().message by LOG(
  *   Implemented ConnectionBuilderManager.
  *   Correct the severity argument for the LOG functions.
  *   Removed all functions owned by ConnectionBuilderManager from Network.
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into ConnectionBuilderManager
  *   Register a sli_logger client for the new Logging infrastructure.
  *   renamed and updated network_impl.h
  *   connect() and register_conn_builder(() are now called from ConnectionBuilderManager
  *   removed network_impl.h
  *   added ConnectionBuilderManager in KernelManager
  *   Format and comments related to Logging infrastructure.
  *   Rough draft of refactoring network -> io_manager
  *   IOManager being worked on
  *   Nightmare of includes
  *   remove extra headers
  *   Compiles & runs
  *   reformat
  *   Remove n_vps_ from communicator.
  *   Merge pull request #1 from apeyser/refactoring_nest
  *   added NodeManager class to nestkernel, moved functions and member variables from Network class to NodeManager, adapted function calls in all scripts
  *   merged remote changes from tammoippen/nest-private refactoring_nest
  *   rm forward declaration in KernelManager
  *   First attempt on NEST API
  *   Put NEST API into nestkernel
  *   Remove warnings of signed vs unsigned int comparissons.
  *   Start ConnGen API
  *   Start librandom API
  *   Adapted function calls to functions that are now in ModelManager
  *   fixed set_status and get_status functions in node_manager.cpp
  *   adapted node_manager.cpp to new logger, fixed function declarations in node_manager.h
  *   fixed Makefile.am
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into modelmanager
  *   Bugfix, formatting
  *   Refer to vp_manager.get_num_threads()
  *   Merge pull request #2 from mdjurfeldt/refactoring_nest
  *   fixed two bugs in node_manager: proper initialization of siblingcontainer and return reference to nodes_vec_ in get_nodes_vec_
  *   Change method signature to lower_case_with_underline
  *   Change method signature to lower_case_with_underline librandom
  *   Merge branch 'refactoring_nest' of https://github.com/tammoippen/nest-private into refactoring_nest
  *   - Created Simulation Manager class
  *   Refactor Conngen datums into separate file.
  *   Manage includes and get correct signatur for conngen API
  *   Merge branch 'refactoring_nest' into simulation_manager
  *   Manage header includes in conngen/
  *   Merge pull request #3 from heplesser/simulation_manager
  *   Rename method names to lower_case_underscore names in conngen API
  *   Merge remote-tracking branch 'refs/remotes/private/refactoring_nest' into refactoring_nest
  *   Format files.
  *   Moved functions and data structures from ConnectionManager to ModelManager
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into modelmanager
  *   Add a topology API
  *   Merge branch 'modelmanager' of github.com:suku248/nest-private into susi
  *   readded set_status and get_status for network class
  *   worked on set_status and get_status
  *   Moved functions from Network to SimulationManager.
  *   Adapt code to move of methods from Network to SimulationManager.
  *   merge remote changes from tammoippen/refactoring_nest
  *   Various changes improving the compilability of your branch.
  *   finished merge
  *   Merge pull request #1 from tammoippen/susi
  *   Initial commit for EventDeliveryManager
  *   Copy EDM methods from Network and replace all occurances of those functions by EDM equivalents
  *   Renamed modelrangemanager in accordance with the manager naming scheme
  *   get_contiguous_gid_range() and get_model_id_of_gid() extracted from
  *   Replace conngen calls; remove nest:: inside namespace
  *   rm recursive definition of get_model_id; add get_model_of_gid; replace calls to that
  *   Merge branch event_delivery into 'refactoring_nest'
  *   Fix wrong header.
  *   Some name-mangeling in topology.
  *   Simulation manager branch now compiles and passes all 735 tests.
  *   Merge branch 'refactoring_nest' into simulation_manager
  *   Removed TimeModifier class.
  *   replaced network_impl.h by connection_builder_manager_impl.h
  *   included connection_builder_manager in kernel_manager
  *   include nest.h instead of nest_types.h
  *   calling functions of the connection_manager from network as a workaround
  *   Merge pull request #5 from heplesser/simulation_manager
  *   Merge branch 'ConnectionBuilderManager' of github.com:hannahbos/nest-private into hannah
  *   Merge branch 'refactoring_nest' of github.com:tammoippen/nest-private into modelmanager
  *   Make ConnBuilderManager compile&run
  *   Comments and renaming according to coding conventions.
  *   Merge branch 'refactor_nodemanager' of github.com:mschmidt87/nest-private into node_manager
  *   Merged mpi_manager from abigailm
  *   Moved remaining functiions to ModelManager and fixed occurences. Code compiles, but gives linker errors.
  *   Merge branch 'modelmanager' of github.com:suku248/nest-private into modelmanager
  *   Forgot to add model_manager_impl.h file
  *   Add comments
  *   Fix up documentation ion io-manager
  *   Initial step to refactor delay from connetormodel
  *   Merge pull request #6 from apeyser/refactoring_nest
  *   Merge branch 'merge_mpi_manager' of github.com:mdjurfeldt/nest-private into mpi_stuff
  *   Fix some leftovers from merging to make it compile
  *   Merge branch 'refactoring_nest' into refactor_delays
  *   make counting connections and delay checking two different topics
  *   only one delaty_checker per thread
  *   Make delay checker not responsible for checking default delau stuff.
  *   make connectormodel responsible for default delay checking
  *   Update Delay Checking and connection counting
  *   Move Time status getting to Simulation manager
  *   Correct return type of freeze/enable_delay_update
  *   Bugfix: min/max_delay not set (correctly) if res changed
  *   Format.
  *   Initial version of RNGManager
  *   put kernel help in kernelmanager
  *   Fix rng manager
  *   Remove Range::OLD_TICS_PER_STEP
  *   Merge commit '8d52babee3780da3a87da380815a929abbe1b0d8' into node_manager
  *   Code now compiles again after merge, but fails most tests. This is most likely due to problems with initializatoin and kernel reset.
  *   Fixed error message on bad delay.
  *   Improved initialization order.
  *   Corrected siblingcontainer_model_ initialization.
  *   Fixed set_status for nodes.
  *   Added TODO comment.
  *   Added missing moduli-initialization to SimulationManager::prepare_simulation().
  *   Renamed manager init and reset to initialize and finalize.
  *   Reordered initialization and finalization of individual managers.
  *   Fixed reset bugs and minor bug.
  *   Tidied up code.
  *   Fix DC copy constructor bug
  *   Some cleanup and refactoring for easier reading
  *   Use of getter function in VPManager
  *   manage some includes
  *   comments
  *   Check that NEST also compiles w/o OMP
  *   Let the python test check, whether nest is actually multi-threaded...
  *   have PMA init in ConnectionBuildeManager
  *   Move PMA initialization in connection_builder_manager
  *   reduce compiler warnings
  *   use correct init function
  *   some formatting
  *   Merge branch 'modelmanager' of github.com:suku248/nest-private into model_manager
  *   Make MM compile
  *   Make MM compile II
  *   put model_manager in kernel_manager functions
  *   remove connection_manager from network and ..
  *   Organize member variables and initialization
  *   Use dictionarys in MM
  *   All DictionaryDatum need a Dictionary to care about
  *   After copy, defaults have to be set to correct id
  *   Copy num_connections for syn_id back at correct place
  *   get default logging output, if nobody registers
  *   Implement reset when threads have changed
  *   Improve MM init and finalize
  *   move dict_miss_is_error into kernel_manager
  *   Refactor dict_miss_checking
  *   Remove using std::vector and that alike
  *   Move memory_info to model_manager
  *   Using kernel manager instead of network
  *   first set of Include Wars
  *   Second set of Include Wars
  *   Includes war part III
  *   Officially add include_checker.py
  *   Cleanup formatting and warnings
  *   Remove the useless functions quit_by_error && get_exitcode
  *   Make publicly used dictionaries DictionaryDatums
  *   Fix bug: double freeing in nestshutdown:
  *   Cleanup Network even more
  *   put execute_sli_protected into sli_neuron
  *   Repopulate ConnectionManager and reset scheduler
  *   Do not use kernel().xyz from xyz
  *   have kernelmanager ask for initialized
  *   Adjust method signatures...
  *   Group associated methods, rm unused methods, formatting
  *   Added music_manager
  *   Finalized music_manager
  *   Gutted communicator.h, communicator_impl.h and communicator.cpp
  *   Organize headers
  *   Merge pull request #7 from mdjurfeldt/communicator
  *   Remove Communicator and Network from build
  *   Still some leftovers from sli-neuron in models/
  *   Repopulate Network.* and Communicator.* for easier future merges
  *   Set Scheduler/Network/ConnectionManager/Communicator back to the NEST 2.8.0 state
  *   Merge branch 'rm_network' into refactoring_nest
  *   Format
  *   rename MPIManager::get_initialized to is_mpi_used
  *   remove friend as far as possible, remove MPIManager::init_
  *   Some cleanup of the MPIManager
  *   Do not use kernel().mpi_manager. inside MPIManager.
  *   fix
  *   fix network size status
  *   make examples run through
  *   get runtime of ./examples/run_exampls.sh
  *   Merge commit 'ed3cab67' into nest3
  *   Adjust to new API
  *   Merge commit 'dc21fdc1' into nest3
  *   Merge commit '235867ea' into nest3 (PR #138)
  *   Merge commit 'dc650c36' into nest3
  *   Merge commit 'c0947a45' into nest3
  *   Merge commit 'c80a8bed' into nest3
  *   Merge commit '164879d8' into nest3
  *   Merge commit 'cc1774c5' into nest3
  *   Merge commit '5845f617' into nest3
  *   Merge commit '5c6535f3' into nest3
  *   Adjust to new API
  *   Merge commit 'ce63c554' into nest3
  *   Merge commit 'c3adb56a' into nest3
  *   Merge commit '349c7286' into nest3 (PR #127)
  *   Merge commit 'ca0513f8' into nest3 (PR #146)
  *   Merge commit 'f2bf8b00' into nest3 (PR #122)
  *   adjust to new api and assign functions to managers
  *   Merge commit 'd9710057' into nest3 (PR #144)
  *   Merge commit 'ecc14fc4' into nest3 (PR #155)
  *   Merge commit 'd6abae69' into nest3 (PR #148)
  *   Merge commit '69d7e170' into nest3 (PR #120)
  *   Merge commit '068f6673' into nest3 (PR #165)
  *   Merge commit '85278e4c' into nest3 (PR #164)
  *   Merge commit '1cbc6862' into nest3 (PR #168)
  *   Merge commit 'a88d7713' into nest3 (PR #159)
  *   Merge commit '5d123272' into nest3 (PR #170)
  *   Merge commit '42a9f7b7' into nest3 (PR #166)
  *   Merge commit '7a79bcf5' into nest3 (PR #174)
  *   Merge commit 'ba001b36' into nest3 (PR #180)
  *   Merge commit 'df46dd4c' into nest3 (PR #175)
  *   Merge commit '2993f420' into nest3 (PR #128)
  *   Merge commit '9e036694' into nest3 (PR #184)
  *   Merge commit '631be698' into nest3 (PR #158)
  *   adjust to new api
  *   remove warnings
  *   Merge commit '8154e51b' into nest3 (PR #189)
  *   Merge commit 'bc2b707d' into nest3 (PR #183)
  *   Merge commit '327644a9' into nest3 (PR #171)
  *   Adjust to new API
  *   Merge commit '7ed0fd07' into nest3
  *   Merge commit '20c3db49' into nest3
  *   Merge commit 'c3be8995' into nest3
  *   Merge commit '470bf93d' into nest3
  *   Merge commit '61abb1a8' into nest3
  *   Merge commit '77fba860' into nest3 (PR #199)
  *   Merge commit '336e89a4' into nest3 (PR #177)
  *   Merge commit '72620b5e' into nest3 (PR #201)
  *   Merge commit '10e238c6' into nest3 (v2.10)
  *   Format.
  *   Move responsibility of growth curve into SPManager.
  *   remove get_node_by_index from nodemanager

File Changes

  *   M .gitignore<https://github.com/nest/nest-simulator/pull/205/files#diff-0> (1)
  *   M configure.ac<https://github.com/nest/nest-simulator/pull/205/files#diff-1> (6)
  *   M conngen/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-2> (4)
  *   M conngen/cg_connect.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-3> (56)
  *   M conngen/cg_connect.h<https://github.com/nest/nest-simulator/pull/205/files#diff-4> (13)
  *   A conngen/conngen.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-5> (190)
  *   A conngen/conngen.h<https://github.com/nest/nest-simulator/pull/205/files#diff-6> (67)
  *   A conngen/conngendatum.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-7> (25)
  *   A conngen/conngendatum.h<https://github.com/nest/nest-simulator/pull/205/files#diff-8> (45)
  *   M conngen/conngenmodule.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-9> (127)
  *   M conngen/conngenmodule.h<https://github.com/nest/nest-simulator/pull/205/files#diff-10> (40)
  *   M examples/MyModule/drop_odd_spike_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-11> (1)
  *   M examples/MyModule/mymodule.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-12> (44)
  *   M examples/MyModule/mymodule.h<https://github.com/nest/nest-simulator/pull/205/files#diff-13> (6)
  *   M examples/MyModule/pif_psc_alpha.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-14> (32)
  *   M examples/MyModule/pif_psc_alpha.h<https://github.com/nest/nest-simulator/pull/205/files#diff-15> (9)
  *   M examples/nest/Brette_et_al_2007/cuba_stdp.sli<https://github.com/nest/nest-simulator/pull/205/files#diff-16> (2)
  *   M examples/run_examples.sh<https://github.com/nest/nest-simulator/pull/205/files#diff-17> (4)
  *   A extras/include_checker.py<https://github.com/nest/nest-simulator/pull/205/files#diff-18> (198)
  *   M libnestutil/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-19> (7)
  *   M libnestutil/allocator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-20> (4)
  *   M libnestutil/allocator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-21> (2)
  *   M libnestutil/compose.hpp<https://github.com/nest/nest-simulator/pull/205/files#diff-22> (5)
  *   M libnestutil/instance.h<https://github.com/nest/nest-simulator/pull/205/files#diff-23> (6)
  *   M libnestutil/lockptr.h<https://github.com/nest/nest-simulator/pull/205/files#diff-24> (1)
  *   A libnestutil/logging.h<https://github.com/nest/nest-simulator/pull/205/files#diff-25> (67)
  *   A libnestutil/logging_event.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-26> (90)
  *   A libnestutil/logging_event.h<https://github.com/nest/nest-simulator/pull/205/files#diff-27> (59)
  *   A libnestutil/logging_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-28> (182)
  *   A libnestutil/logging_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-29> (126)
  *   A libnestutil/manager_interface.h<https://github.com/nest/nest-simulator/pull/205/files#diff-30> (95)
  *   M libnestutil/numerics.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-31> (1)
  *   M libnestutil/numerics.h<https://github.com/nest/nest-simulator/pull/205/files#diff-32> (6)
  *   M libnestutil/propagator_stability.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-33> (4)
  *   M libnestutil/stopwatch.h<https://github.com/nest/nest-simulator/pull/205/files#diff-34> (5)
  *   M librandom/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-35> (3)
  *   M librandom/binomial_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-36> (12)
  *   M librandom/binomial_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-37> (16)
  *   M librandom/clipped_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-38> (11)
  *   M librandom/exp_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-39> (5)
  *   M librandom/exp_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-40> (9)
  *   M librandom/gamma_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-41> (7)
  *   M librandom/gamma_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-42> (9)
  *   M librandom/gsl_binomial_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-43> (13)
  *   M librandom/gsl_binomial_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-44> (15)
  *   M librandom/gslrandomgen.h<https://github.com/nest/nest-simulator/pull/205/files#diff-45> (12)
  *   M librandom/knuthlfg.h<https://github.com/nest/nest-simulator/pull/205/files#diff-46> (5)
  *   M librandom/librandom_exceptions.h<https://github.com/nest/nest-simulator/pull/205/files#diff-47> (1)
  *   M librandom/lognormal_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-48> (10)
  *   M librandom/lognormal_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-49> (9)
  *   M librandom/mt19937.h<https://github.com/nest/nest-simulator/pull/205/files#diff-50> (5)
  *   M librandom/normal_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-51> (10)
  *   M librandom/normal_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-52> (9)
  *   M librandom/poisson_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-53> (13)
  *   M librandom/poisson_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-54> (8)
  *   A librandom/random.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-55> (120)
  *   A librandom/random.h<https://github.com/nest/nest-simulator/pull/205/files#diff-56> (51)
  *   M librandom/random_datums.h<https://github.com/nest/nest-simulator/pull/205/files#diff-57> (8)
  *   M librandom/random_numbers.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-58> (79)
  *   M librandom/random_numbers.h<https://github.com/nest/nest-simulator/pull/205/files#diff-59> (14)
  *   M librandom/randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-60> (11)
  *   M librandom/randomgen.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-61> (2)
  *   M librandom/randomgen.h<https://github.com/nest/nest-simulator/pull/205/files#diff-62> (10)
  *   M librandom/randomtest.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-63> (37)
  *   M librandom/uniform_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-64> (7)
  *   M librandom/uniform_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-65> (9)
  *   M librandom/uniformint_randomdev.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-66> (11)
  *   M librandom/uniformint_randomdev.h<https://github.com/nest/nest-simulator/pull/205/files#diff-67> (9)
  *   M models/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-68> (1)
  *   M models/ac_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-69> (23)
  *   M models/ac_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-70> (8)
  *   M models/aeif_cond_alpha.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-71> (42)
  *   M models/aeif_cond_alpha.h<https://github.com/nest/nest-simulator/pull/205/files#diff-72> (19)
  *   M models/aeif_cond_alpha_RK5.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-73> (42)
  *   M models/aeif_cond_alpha_RK5.h<https://github.com/nest/nest-simulator/pull/205/files#diff-74> (7)
  *   M models/aeif_cond_alpha_multisynapse.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-75> (32)
  *   M models/aeif_cond_alpha_multisynapse.h<https://github.com/nest/nest-simulator/pull/205/files#diff-76> (9)
  *   M models/aeif_cond_exp.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-77> (42)
  *   M models/aeif_cond_exp.h<https://github.com/nest/nest-simulator/pull/205/files#diff-78> (19)
  *   M models/amat2_psc_exp.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-79> (33)
  *   M models/amat2_psc_exp.h<https://github.com/nest/nest-simulator/pull/205/files#diff-80> (12)
  *   M models/binary_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-81> (20)
  *   M models/binary_neuron_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-82> (40)
  *   M models/cont_delay_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-83> (4)
  *   M models/cont_delay_connection_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-84> (16)
  *   M models/correlation_detector.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-85> (14)
  *   M models/correlation_detector.h<https://github.com/nest/nest-simulator/pull/205/files#diff-86> (10)
  *   M models/correlomatrix_detector.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-87> (19)
  *   M models/correlomatrix_detector.h<https://github.com/nest/nest-simulator/pull/205/files#diff-88> (10)
  *   M models/correlospinmatrix_detector.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-89> (19)
  *   M models/correlospinmatrix_detector.h<https://github.com/nest/nest-simulator/pull/205/files#diff-90> (10)
  *   M models/dc_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-91> (13)
  *   M models/dc_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-92> (7)
  *   M models/gamma_sup_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-93> (29)
  *   M models/gamma_sup_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-94> (14)
  *   M models/ginzburg_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-95> (1)
  *   M models/hh_cond_exp_traub.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-96> (43)
  *   M models/hh_cond_exp_traub.h<https://github.com/nest/nest-simulator/pull/205/files#diff-97> (15)
  *   M models/hh_psc_alpha.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-98> (40)
  *   M models/hh_psc_alpha.h<https://github.com/nest/nest-simulator/pull/205/files#diff-99> (23)
  *   M models/hh_psc_alpha_gap.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-100> (60)
  *   M models/hh_psc_alpha_gap.h<https://github.com/nest/nest-simulator/pull/205/files#diff-101> (22)
  *   M models/ht_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-102> (1)
  *   M models/ht_neuron.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-103> (16)
  *   M models/ht_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-104> (24)
  *   M models/iaf_chs_2007.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-105> (28)
  *   M models/iaf_chs_2007.h<https://github.com/nest/nest-simulator/pull/205/files#diff-106> (15)
  *   M models/iaf_chxk_2008.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-107> (39)
  *   M models/iaf_chxk_2008.h<https://github.com/nest/nest-simulator/pull/205/files#diff-108> (19)
  *   M models/iaf_cond_alpha.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-109> (39)
  *   M models/iaf_cond_alpha.h<https://github.com/nest/nest-simulator/pull/205/files#diff-110> (19)
  *   M models/iaf_cond_alpha_mc.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-111> (36)
  *   M models/iaf_cond_alpha_mc.h<https://github.com/nest/nest-simulator/pull/205/files#diff-112> (26)
  *   M models/iaf_cond_exp.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-113> (40)
  *   M models/iaf_cond_exp.h<https://github.com/nest/nest-simulator/pull/205/files#diff-114> (19)
  *   M models/iaf_cond_exp_sfa_rr.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-115> (39)
  *   M models/iaf_cond_exp_sfa_rr.h<https://github.com/nest/nest-simulator/pull/205/files#diff-116> (22)
  *   M models/iaf_neuron.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-117> (33)
  *   M models/iaf_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-118> (9)
  *   M models/iaf_psc_alpha.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-119> (36)
  *   M models/iaf_psc_alpha.h<https://github.com/nest/nest-simulator/pull/205/files#diff-120> (11)
  *   M models/iaf_psc_alpha_multisynapse.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-121> (33)
  *   M models/iaf_psc_alpha_multisynapse.h<https://github.com/nest/nest-simulator/pull/205/files#diff-122> (11)
  *   M models/iaf_psc_delta.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-123> (31)
  *   M models/iaf_psc_delta.h<https://github.com/nest/nest-simulator/pull/205/files#diff-124> (10)
  *   M models/iaf_psc_exp.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-125> (40)
  *   M models/iaf_psc_exp.h<https://github.com/nest/nest-simulator/pull/205/files#diff-126> (11)
  *   M models/iaf_psc_exp_multisynapse.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-127> (33)
  *   M models/iaf_psc_exp_multisynapse.h<https://github.com/nest/nest-simulator/pull/205/files#diff-128> (11)
  *   M models/iaf_tum_2000.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-129> (36)
  *   M models/iaf_tum_2000.h<https://github.com/nest/nest-simulator/pull/205/files#diff-130> (11)
  *   M models/izhikevich.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-131> (34)
  *   M models/izhikevich.h<https://github.com/nest/nest-simulator/pull/205/files#diff-132> (10)
  *   M models/mat2_psc_exp.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-133> (33)
  *   M models/mat2_psc_exp.h<https://github.com/nest/nest-simulator/pull/205/files#diff-134> (12)
  *   M models/mcculloch_pitts_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-135> (1)
  *   M models/mip_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-136> (21)
  *   M models/mip_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-137> (9)
  *   M models/modelsmodule.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-138> (324)
  *   M models/modelsmodule.h<https://github.com/nest/nest-simulator/pull/205/files#diff-139> (13)
  *   M models/multimeter.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-140> (6)
  *   M models/multimeter.h<https://github.com/nest/nest-simulator/pull/205/files#diff-141> (18)
  *   M models/music_cont_in_proxy.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-142> (22)
  *   M models/music_cont_in_proxy.h<https://github.com/nest/nest-simulator/pull/205/files#diff-143> (19)
  *   M models/music_event_in_proxy.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-144> (34)
  *   M models/music_event_in_proxy.h<https://github.com/nest/nest-simulator/pull/205/files#diff-145> (10)
  *   M models/music_event_out_proxy.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-146> (25)
  *   M models/music_event_out_proxy.h<https://github.com/nest/nest-simulator/pull/205/files#diff-147> (13)
  *   M models/music_message_in_proxy.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-148> (23)
  *   M models/music_message_in_proxy.h<https://github.com/nest/nest-simulator/pull/205/files#diff-149> (21)
  *   M models/noise_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-150> (24)
  *   M models/noise_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-151> (12)
  *   M models/parrot_neuron.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-152> (29)
  *   M models/parrot_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-153> (9)
  *   M models/poisson_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-154> (17)
  *   M models/poisson_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-155> (11)
  *   M models/pp_pop_psc_delta.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-156> (39)
  *   M models/pp_pop_psc_delta.h<https://github.com/nest/nest-simulator/pull/205/files#diff-157> (13)
  *   M models/pp_psc_delta.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-158> (36)
  *   M models/pp_psc_delta.h<https://github.com/nest/nest-simulator/pull/205/files#diff-159> (16)
  *   M models/ppd_sup_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-160> (29)
  *   M models/ppd_sup_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-161> (14)
  *   M models/pulsepacket_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-162> (31)
  *   M models/pulsepacket_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-163> (13)
  *   M models/quantal_stp_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-164> (11)
  *   M models/quantal_stp_connection_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-165> (5)
  *   M models/sinusoidal_gamma_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-166> (41)
  *   M models/sinusoidal_gamma_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-167> (16)
  *   M models/sinusoidal_poisson_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-168> (39)
  *   M models/sinusoidal_poisson_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-169> (12)
  *   M models/spike_detector.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-170> (54)
  *   M models/spike_detector.h<https://github.com/nest/nest-simulator/pull/205/files#diff-171> (18)
  *   M models/spike_dilutor.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-172> (24)
  *   M models/spike_dilutor.h<https://github.com/nest/nest-simulator/pull/205/files#diff-173> (7)
  *   M models/spike_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-174> (44)
  *   M models/spike_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-175> (36)
  *   M models/spin_detector.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-176> (54)
  *   M models/spin_detector.h<https://github.com/nest/nest-simulator/pull/205/files#diff-177> (18)
  *   M models/static_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-178> (1)
  *   M models/static_connection_hom_w.h<https://github.com/nest/nest-simulator/pull/205/files#diff-179> (3)
  *   M models/stdp_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-180> (13)
  *   M models/stdp_connection_facetshw_hom.h<https://github.com/nest/nest-simulator/pull/205/files#diff-181> (7)
  *   M models/stdp_connection_facetshw_hom_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-182> (11)
  *   M models/stdp_connection_hom.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-183> (11)
  *   M models/stdp_connection_hom.h<https://github.com/nest/nest-simulator/pull/205/files#diff-184> (5)
  *   M models/stdp_dopa_connection.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-185> (15)
  *   M models/stdp_dopa_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-186> (23)
  *   M models/stdp_pl_connection_hom.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-187> (12)
  *   M models/stdp_pl_connection_hom.h<https://github.com/nest/nest-simulator/pull/205/files#diff-188> (6)
  *   M models/step_current_generator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-189> (13)
  *   M models/step_current_generator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-190> (7)
  *   M models/tsodyks2_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-191> (6)
  *   M models/tsodyks_connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-192> (4)
  *   M models/tsodyks_connection_hom.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-193> (3)
  *   M models/tsodyks_connection_hom.h<https://github.com/nest/nest-simulator/pull/205/files#diff-194> (4)
  *   M models/volume_transmitter.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-195> (40)
  *   M models/volume_transmitter.h<https://github.com/nest/nest-simulator/pull/205/files#diff-196> (16)
  *   M nest/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-197> (3)
  *   M nest/main.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-198> (17)
  *   M nest/neststartup.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-199> (114)
  *   M nest/neststartup.h<https://github.com/nest/nest-simulator/pull/205/files#diff-200> (35)
  *   R nest/sli_neuron.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-201> (90)
  *   R nest/sli_neuron.h<https://github.com/nest/nest-simulator/pull/205/files#diff-202> (17)
  *   M nestkernel/Makefile.am<https://github.com/nest/nest-simulator/pull/205/files#diff-203> (28)
  *   M nestkernel/archiving_node.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-204> (4)
  *   M nestkernel/archiving_node.h<https://github.com/nest/nest-simulator/pull/205/files#diff-205> (17)
  *   M nestkernel/common_properties_hom_w.h<https://github.com/nest/nest-simulator/pull/205/files#diff-206> (2)
  *   M nestkernel/common_synapse_properties.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-207> (8)
  *   M nestkernel/common_synapse_properties.h<https://github.com/nest/nest-simulator/pull/205/files#diff-208> (9)
  *   M nestkernel/communicator.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-209> (4)
  *   M nestkernel/communicator.h<https://github.com/nest/nest-simulator/pull/205/files#diff-210> (41)
  *   M nestkernel/communicator_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-211> (4)
  *   M nestkernel/conn_builder.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-212> (0)
  *   I nestkernel/conn_builder.h<https://github.com/nest/nest-simulator/pull/205/files#diff-213> (0)
  *   I nestkernel/conn_builder_factory.h<https://github.com/nest/nest-simulator/pull/205/files#diff-214> (0)
  *   I nestkernel/conn_parameter.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-215> (0)
  *   I nestkernel/conn_parameter.h<https://github.com/nest/nest-simulator/pull/205/files#diff-216> (0)
  *   I nestkernel/connection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-217> (0)
  *   A nestkernel/connection_builder_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-218> (0)
  *   A nestkernel/connection_builder_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-219> (0)
  *   A nestkernel/connection_builder_manager_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-220> (0)
  *   I nestkernel/connection_id.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-221> (0)
  *   I nestkernel/connection_id.h<https://github.com/nest/nest-simulator/pull/205/files#diff-222> (0)
  *   I nestkernel/connection_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-223> (0)
  *   I nestkernel/connection_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-224> (0)
  *   I nestkernel/connector_base.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-225> (0)
  *   I nestkernel/connector_base.h<https://github.com/nest/nest-simulator/pull/205/files#diff-226> (0)
  *   I nestkernel/connector_model.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-227> (0)
  *   I nestkernel/connector_model.h<https://github.com/nest/nest-simulator/pull/205/files#diff-228> (0)
  *   I nestkernel/connector_model_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-229> (0)
  *   A nestkernel/delay_checker.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-230> (0)
  *   A nestkernel/delay_checker.h<https://github.com/nest/nest-simulator/pull/205/files#diff-231> (0)
  *   I nestkernel/device.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-232> (0)
  *   I nestkernel/device.h<https://github.com/nest/nest-simulator/pull/205/files#diff-233> (0)
  *   I nestkernel/dynamicloader.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-234> (0)
  *   I nestkernel/dynamicloader.h<https://github.com/nest/nest-simulator/pull/205/files#diff-235> (0)
  *   I nestkernel/event.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-236> (0)
  *   I nestkernel/event.h<https://github.com/nest/nest-simulator/pull/205/files#diff-237> (0)
  *   A nestkernel/event_delivery_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-238> (0)
  *   A nestkernel/event_delivery_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-239> (0)
  *   A nestkernel/event_delivery_manager_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-240> (0)
  *   I nestkernel/event_priority.h<https://github.com/nest/nest-simulator/pull/205/files#diff-241> (0)
  *   I nestkernel/exceptions.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-242> (0)
  *   I nestkernel/exceptions.h<https://github.com/nest/nest-simulator/pull/205/files#diff-243> (0)
  *   I nestkernel/genericmodel.h<https://github.com/nest/nest-simulator/pull/205/files#diff-244> (0)
  *   I nestkernel/gid_collection.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-245> (0)
  *   I nestkernel/gid_collection.h<https://github.com/nest/nest-simulator/pull/205/files#diff-246> (0)
  *   I nestkernel/growth_curve.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-247> (0)
  *   I nestkernel/growth_curve.h<https://github.com/nest/nest-simulator/pull/205/files#diff-248> (0)
  *   I nestkernel/growth_curve_factory.h<https://github.com/nest/nest-simulator/pull/205/files#diff-249> (0)
  *   I nestkernel/histentry.h<https://github.com/nest/nest-simulator/pull/205/files#diff-250> (0)
  *   A nestkernel/io_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-251> (0)
  *   A nestkernel/io_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-252> (0)
  *   A nestkernel/kernel_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-253> (0)
  *   A nestkernel/kernel_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-254> (0)
  *   I nestkernel/model.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-255> (0)
  *   I nestkernel/model.h<https://github.com/nest/nest-simulator/pull/205/files#diff-256> (0)
  *   A nestkernel/model_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-257> (0)
  *   A nestkernel/model_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-258> (0)
  *   A nestkernel/model_manager_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-259> (0)
  *   I nestkernel/modelrange.h<https://github.com/nest/nest-simulator/pull/205/files#diff-260> (0)
  *   A nestkernel/modelrange_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-261> (0)
  *   A nestkernel/modelrange_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-262> (0)
  *   D nestkernel/modelrangemanager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-263> (0)
  *   D nestkernel/modelrangemanager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-264> (0)
  *   A nestkernel/mpi_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-265> (0)
  *   A nestkernel/mpi_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-266> (0)
  *   A nestkernel/mpi_manager_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-267> (0)
  *   I nestkernel/multirange.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-268> (0)
  *   I nestkernel/multirange.h<https://github.com/nest/nest-simulator/pull/205/files#diff-269> (0)
  *   I nestkernel/music_event_handler.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-270> (0)
  *   I nestkernel/music_event_handler.h<https://github.com/nest/nest-simulator/pull/205/files#diff-271> (0)
  *   A nestkernel/music_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-272> (0)
  *   A nestkernel/music_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-273> (0)
  *   A nestkernel/nest.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-274> (0)
  *   I nestkernel/nest.h<https://github.com/nest/nest-simulator/pull/205/files#diff-275> (0)
  *   I nestkernel/nest_datums.h<https://github.com/nest/nest-simulator/pull/205/files#diff-276> (0)
  *   I nestkernel/nest_names.h<https://github.com/nest/nest-simulator/pull/205/files#diff-277> (0)
  *   I nestkernel/nest_time.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-278> (0)
  *   I nestkernel/nest_time.h<https://github.com/nest/nest-simulator/pull/205/files#diff-279> (0)
  *   I nestkernel/nest_timeconverter.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-280> (0)
  *   I nestkernel/nest_timeconverter.h<https://github.com/nest/nest-simulator/pull/205/files#diff-281> (0)
  *   D nestkernel/nest_timemodifier.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-282> (0)
  *   D nestkernel/nest_timemodifier.h<https://github.com/nest/nest-simulator/pull/205/files#diff-283> (0)
  *   A nestkernel/nest_types.h<https://github.com/nest/nest-simulator/pull/205/files#diff-284> (0)
  *   I nestkernel/nestmodule.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-285> (0)
  *   I nestkernel/nestmodule.h<https://github.com/nest/nest-simulator/pull/205/files#diff-286> (0)
  *   I nestkernel/network.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-287> (0)
  *   I nestkernel/network.h<https://github.com/nest/nest-simulator/pull/205/files#diff-288> (0)
  *   D nestkernel/network_impl.h<https://github.com/nest/nest-simulator/pull/205/files#diff-289> (0)
  *   I nestkernel/node.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-290> (0)
  *   I nestkernel/node.h<https://github.com/nest/nest-simulator/pull/205/files#diff-291> (0)
  *   A nestkernel/node_manager.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-292> (0)
  *   A nestkernel/node_manager.h<https://github.com/nest/nest-simulator/pull/205/files#diff-293> (0)
  *   I nestkernel/nodelist.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-294> (0)
  *   I nestkernel/nodelist.h<https://github.com/nest/nest-simulator/pull/205/files#diff-295> (0)
  *   I nestkernel/proxynode.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-296> (0)
  *   I nestkernel/proxynode.h<https://github.com/nest/nest-simulator/pull/205/files#diff-297> (0)
  *   I nestkernel/pseudo_recording_device.h<https://github.com/nest/nest-simulator/pull/205/files#diff-298> (0)
  *   I nestkernel/recordables_map.h<https://github.com/nest/nest-simulator/pull/205/files#diff-299> (0)
  *   I nestkernel/recording_device.cpp<https://github.com/nest/nest-simulator/pull/205/files#diff-300> (0)

Patch Links:

  *   https://github.com/nest/nest-simulator/pull/205.patch
  *   https://github.com/nest/nest-simulator/pull/205.diff

—
Reply to this email directly or view it on GitHub<https://github.com/nest/nest-simulator/pull/205>.

apeyser commented:
Would this be worth rebasing onto the master?

janhahne commented:
@tammoippen The tests from the testsuite are sufficient to confirm that the gap junctions work fine. From looking at the source code I can confirm that you missed nothing and that all of the later applied changes to the gap framework are present. Very good work!

janhahne commented:
@tammoippen I just noticed a small influence of this PR on the gap junction framework. The delay of gap_junction connections is no longer ignored despite of how they are created ( see issue #208 ). In 2.10.0 it worked at least for one case.

sdiazpier commented:
Dear @tammoippen I have tested all the structural plasticity code and it is working well. I have also checked all the related classes and everything seems to be in place for the structural plasticity functionality.

jakobj commented on nestkernel/Makefile.am:
```diff
@@ -54,25 +52,23 @@ libnest_la_SOURCES=\
 		histentry.h histentry.cpp\
 		model.h model.cpp\
-		nest.h\
+		model_manager.h model_manager.cpp\
```
model_manager_impl.h is missing?

jakobj commented on nestkernel/Makefile.am:
```diff
@@ -83,7 +79,21 @@ libnest_la_SOURCES=\
+		io_manager.h io_manager.cpp \
+		mpi_manager.h mpi_manager_impl.h mpi_manager.cpp \
+		simulation_manager.h simulation_manager.cpp \
+		connection_builder_manager.h connection_builder_manager.cpp \
```
connection_builder_manager_impl.h is missing?

tammoippen commented:
Hi @nest/nest-developers,

we plan to merge the PR by the end of Jan. Please have a look into the changes again and find errors or improvements. I added some graphics to help you fiddle your way through the managers:

[ManagerRelations.pdf](https://github.com/nest/nest-simulator/files/103128/ManagerRelations.pdf)

[ManagerDetails.pdf](https://github.com/nest/nest-simulator/files/103129/ManagerDetails.pdf)

Thank you! :smile:

jakobj commented:
i think you might be missing the include guards in `model_manager_impl.h` and `universal_data_logger.h`

tammoippen commented:
@jakobj Thanks for catching this. Actually, there were a few other `*_impl.h` missing include guards.

heplesser commented:
I just compiled NEST3 with `clang-omp` from Homebrew under OSX 10.10. It spat out a number of warnings, and some tests fail.

@tammoippen How should I best contribute small fixes, such as missing newlines at end of file?

@sdiazpier Clang points out (correctly) that `SPBuilder::connect(GIDCollection, GIDCollection)` shadows `ConnBuilder::connect()`. Both appear to be rather different functions, since one takes two arguments and the other none. The two-argument version is apparently only called from `SPManager::create_synapses()`. So I think the function in `SPBuilder` should probably get a different name. A positive side effect would be that it is no longer virtual, making it (likely) faster.

@sdiazpier Clang also issues the following warning:
```
../../src/nestkernel/sp_manager.cpp:622:21: warning: comparison of integers of different signs: 'int' and 'size_type' (aka 'unsigned long') [-Wsign-compare]
    if ( -( *n_it ) > global_sources.size() )
```

@jougs Clang warns about a few deprecated MPI-2 names:
```
../../src/nestkernel/mpi_manager.cpp:116:3: warning: 'MPI_Address' is deprecated: MPI_Address is superseded by MPI_Get_address in MPI-2.0 [-Wdeprecated-declarations]
  MPI_Address( &( ogs.offset_ ), &address );
  ^
/usr/local/include/mpi.h:1187:20: note: 'MPI_Address' has been explicitly marked deprecated here
OMPI_DECLSPEC  int MPI_Address(void *location, MPI_Aint *address)
                   ^
../../src/nestkernel/mpi_manager.cpp:122:3: warning: 'MPI_Type_struct' is deprecated: MPI_Type_struct is superseded by MPI_Type_create_struct in MPI-2.0
      [-Wdeprecated-declarations]
  MPI_Type_struct( 2, blockcounts, offsets, source_types, &MPI_OFFGRID_SPIKE );
```

Unfortunately, several tests fail:
- Running test 'unittests/test_aeif_cond_alpha_multisynapse.sli'... Failed: segmentation fault
- Running test 'unittests/test_mip_corrdet.sli'... Failed: missed C++ assertion
- Running test 'unittests/test_multithreading_devices.sli'... Failed: segmentation fault
- Running test 'unittests/test_recorder_close_flush.sli'... Failed: missed SLI assertion

Finally the testsuite hangs forever on Running test 'regressiontests/ticket-80-175-179.sli'...
No tests later in the testsuite are run.

I haven't had time to investigate.

tammoippen commented:
> How should I best contribute small fixes, such as missing newlines at end of file?
Please create a PR onto my Fork. I will merge ASAP.

> Unfortunately, several tests fail:
> 
> Running test 'unittests/test_aeif_cond_alpha_multisynapse.sli'... Failed: segmentation fault
> Running test 'unittests/test_mip_corrdet.sli'... Failed: missed C++ assertion
> Running test 'unittests/test_multithreading_devices.sli'... Failed: segmentation fault
> Running test 'unittests/test_recorder_close_flush.sli'... Failed: missed SLI assertion
> Finally the testsuite hangs forever on Running test 'regressiontests/ticket-80-175-179.sli'...
> No tests later in the testsuite are run.
> 
Some of these tests also fail for NEST 2.10.0 on OS X with clang. I think there is a specific issue with this combination (on Linux with clang I did not observe any of these issues).

heplesser commented:
I have explored the test failures a bit further (all with `clang-omp`):

- `test_aeif_cond_alpha_multisynapse.sli` 
  - fails on most runs with a segfault, usually somewhere during `update()`
  - detailed locations vary
  - on rare occasions, it does not crash, but appears to hang
  - in one case, it seg faulted during `finalize_nodes()`
  - the test uses only a single thread
- `test_mip_corrdet_sli`
  - uses single thread
  - seg faults reliably
  - mostly in `finalize_nodes()`, once in `update()` 
- `test_multithreading_devices.sli`
 - uses 1 or 2 threads
 - seg faults reliably
 - `test_multithreading.sli` passes reliably
- `test_recorder_close_flush.sli`
  - seg faults reliably
  - on one occasion, it ran for a while before segfaulting
- `ticket-80-175-179.sli`
  - seg faults in most cases
  - hangs occasionally

Tests coming after `ticket-80-175-179.sli` have not run yet.

tammoippen commented:
Thank you for finding and resolving all these issues. I addressed the remaining issues in #74 and #212 - I do not think, that they should block this PR. Hence, I am merging this PR and lifting the code freeze.
